### PR TITLE
fix(cache): stop native archive timestamps from missing cached actions

### PR DESCRIPTION
**Pull request by jdx**
https://entire.io/gh/jdx/mr-boxington/trails/15

## The problem

Apple's `ar` and `ranlib` stamp the current time into an archive's symbol table and member headers. A build script that produces a `.a` therefore hands Cargo an input whose bytes change between otherwise identical builds — the archive's digest moves, and **every cached action downstream of it misses**, even though no member payload changed.

CMake-based native dependencies hit this routinely, because they invoke `/usr/bin/ar` directly rather than going through the `cc` crate, which already sets `ZERO_AR_DATE` for implementations it knows to be non-deterministic.

The effect is not subtle. Surveying an existing mbx cache on macOS, `libz-ng.a` appeared under **7 distinct digests across 7 builds** — identical 564,280 bytes each time, all 36 member payloads byte-identical, differing only in the header timestamp. Every crate downstream of it missed on every fresh-target build.

## What changes

mbx now sets `ZERO_AR_DATE` for the build scripts it runs, controlled by a new `ar_determinism` setting:

| value | behavior |
|---|---|
| `auto` (default) | normalize every profile except `release` |
| `always` | normalize `release` too |
| `off` | leave the toolchain alone |

```sh
# default: dev builds get stable archives, release artifacts are untouched
mbx build                          # build scripts see ZERO_AR_DATE=1
mbx build --release                # build scripts see it unset

MBX_AR_DETERMINISM=always mbx build --release   # normalize release too
```

`auto` excludes `release` deliberately: the cache misses this prevents are worst in the edit-build loop, while a release build is the one whose exact output bytes someone may be signing, publishing, or diffing against a previous artifact. A `ZERO_AR_DATE` the developer set themselves is never overridden, under any mode.

The effective value — the `ZERO_AR_DATE` the script will actually read, including an inherited one — is part of the build-script action key, so changing the setting takes effect on the next build instead of whenever the cache happens to miss for some other reason. Two developers with different settings do not share an entry. The action version is bumped alongside it, because an entry stored before this cannot say which policy produced it; existing build-script entries are therefore recomputed once.

`mbx doctor` gains an `archives` check that probes the local tools and reports what it finds:

```
  ok  archives     the archive tools stamp a timestamp; mbx sets ZERO_AR_DATE
                   for build scripts (release left as the toolchain makes it)
```

It warns instead when the tools stamp and the configured policy leaves this build unnormalized, and when the probe could not run at all.

## This fixes one cause, and the PR says which

Timestamp-only drift is what `ZERO_AR_DATE` addresses. It is not the only way an archive's bytes move, and the other case in this cache turned out **not** to be instability at all.

`libaws_lc_0_45_0_crypto.a` also appears under 7 distinct digests. Those copies are byte-identical compilations recorded at different build directories: an object carries the directory it was compiled in, so two checkouts produce different bytes by design. On two copies of the same size the entire difference is **two bytes** — `mise-pr-13386` against `mise-pr-13387` — and all 119 differing members become identical once that path is normalized. Across copies the archive size tracks path length exactly: `8295440 - 8294344 = 1096 = 274 objects x 4 characters`.

mbx already handles this correctly. The `cc` shim passes `-fdebug-prefix-map` and refuses to publish an object still carrying the real path, and a build script whose `OUT_DIR` embeds paths is keyed to that path instead of being shared. Nothing here is broken; these archives simply cannot be reused across checkouts.

`benchmarks/archive_stability.py` surveys an existing cache and separates the cases:

```
$ python3 benchmarks/archive_stability.py
archives comparable across target directories: 28
unstable: 6

  debug/libz-ng-sys-*/out/lib/libz-ng.a
    7 distinct digests from 7 copies; cause: timestamp
    every member payload identical -> ZERO_AR_DATE addresses this

  debug/aws-lc-sys-*/out/build/artifacts/libaws_lc_0_45_0_crypto.a
    7 distinct digests from 7 copies; cause: build path
    120 members differ, and the copies record different build directories
      built at: /Users/jdx/.codex/worktrees/4c81/mise/target/debug/build/...
      built at: /private/tmp/mise-pr-13385/target/debug/build/...
```

It compares every distinct variant in a group, and judges substance from member payloads only — `__.SYMDEF` is the archive's own index, rebuilt from member offsets, so it differs whenever a path length does.

## Validation

`test/archive_determinism.bats` covers the policy matrix (each profile, each mode, and the developer's own setting winning), a policy change taking effect against a warm store, and an end-to-end case: a build script that archives an unchanged object twice, asserting the archive keeps its digest.

The two cases that guard the cache behavior were each confirmed to **fail** when their fix is stubbed out, so they detect the bug rather than merely passing.

`mise run ci` is otherwise green: 610 lib tests and 75 behavioral tests. One pre-existing failure remains — `setup.bats` "mise postinstall activates transparent Cargo in local and global scopes" — which fails identically on the base commit `bbd7e95` and is unrelated to this change.

## Notes

The variable is set on every host rather than only macOS: it is read by Apple's archive tools wherever they run, including cctools cross-compiling to an Apple target from Linux, and is inert for toolchains that do not know it.

The spawn path and the key path both read the variable with `var_os`, and the key spells it as `text:<value>` or `bytes:<hex>`. An inherited value that is not UTF-8 still reaches the build script, so keying it as absent would let that build share an entry with one where mbx supplied its own `1`.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

---

> [!NOTE]
> **Medium Risk**
> Changes build-script environment and cache key versioning; wrong keying or policy could restore mismatched native artifacts, though version bump and tests mitigate that.
> 
> **Overview**
> Build scripts that produce native `.a` archives were changing digest on every run when `ar`/`ranlib` stamped timestamps, which cascaded into cache misses downstream. **mbx now sets `ZERO_AR_DATE=1` for build script runs** (unless the developer already set it), governed by new **`ar_determinism`** (`auto` / `always` / `off`, default `auto` skips `release`).
> 
> The **effective value is included in the build-script cache action key** and the action **version is bumped to 3**, so policy changes apply on the next build instead of replaying archives from an old policy. **`mbx doctor`** adds an **archives** probe that reports whether local tools stamp timestamps and whether the current config normalizes them.
> 
> Adds **`benchmarks/archive_stability.py`** to classify duplicate archives in a target tree (timestamp vs build-path vs content), plus **`test/archive_determinism.bats`** for the policy matrix, warm-cache policy toggles, and stable archive digests when stamping tools are present.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit 4c6e6d08349c5ad2a7f8de25cbb6a5c41ba81294. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

### CODE REVIEW

File: crates/mbx/src/build_script.rs [hunk 1]
```diff
@@ -220,6 +221,24 @@ fn install_launcher(mbx: &Path, executable: &Path) -> std::io::Result<()> {
...
+    let mode = std::env::var(session::AR_DETERMINISM_ENV).map_or_else(
+        |_| ArDeterminism::default(),
+        |value| ArDeterminism::parse(&value),
+    );
+    let profile = std::env::var(crate::ar::PROFILE).ok();
+    let already_set = std::env::var_os(crate::ar::ZERO_AR_DATE).is_some();
+    if crate::ar::normalizes(mode, profile.as_deref(), already_set) {
+        command.env(crate::ar::ZERO_AR_DATE, "1");
+    }
+}
```
[the remark this answers is not in the document]
Comment (jdx) [reply crates/mbx/src/build_script.rs:238]: Fixed in 66a0f21 — the build-script action key now includes the `ZERO_AR_DATE` the script will actually read (including a developer-set value), and the action version is bumped, so a policy change is a different entry rather than a stale restore. Added a bats case that flips the policy against a warm store; it fails without the key change.

_🤖 Addressed by [Claude Code](https://claude.com/claude-code)_

File: crates/mbx/src/doctor.rs [hunk 2]
```diff
@@ -369,6 +370,59 @@ fn reflink_check_with(
...
+    let mode = crate::ar::ArDeterminism::parse(&config.ar_determinism);
+    match crate::ar::probe() {
+        crate::ar::Probe::Deterministic => Check::pass(
+            "archives",
+            "the archive tools already produce identical bytes for identical input",
+        ),
+        crate::ar::Probe::Unavailable(reason) => Check::pass(
+            "archives",
+            format!("archive determinism could not be probed ({reason})"),
+        ),
```
[the remark this answers is not in the document]
Comment (jdx) [reply crates/mbx/src/doctor.rs]: Fixed in 66a0f21 — an unavailable probe is now `Check::warn` rather than `pass`, matching the other probes here; a green line read as "archives are fine" when nothing had been assessed.

_🤖 Addressed by [Claude Code](https://claude.com/claude-code)_

File: benchmarks/archive_stability.py [hunk 3]
```diff
@@ -0,0 +1,175 @@
...
+    representatives = [group[0] for group in by_digest.values()]
+    sizes = sorted({path.stat().st_size for path in paths})
+    first, second = members(representatives[0]), members(representatives[1])
+    if first is None or second is None:
+        cause, differing = "unknown", None
+    elif set(first) != set(second):
+        cause, differing = "content", None
+    else:
+        differing = sorted(name for name in first if first[name] != second[name])
+        cause = "content" if differing else "timestamp"
```
[the remark this answers is not in the document]
Comment (jdx) [reply benchmarks/archive_stability.py]: Fixed in 66a0f21 — `classify` now compares every digest variant against the baseline and unions the differing members, so `timestamp` means all copies agree on every payload. On the sample cache this reclassifies nothing but widens `libaws_lc_crypto.a` from 119 to 120 differing members, catching `__.SYMDEF SORTED`.

_🤖 Addressed by [Claude Code](https://claude.com/claude-code)_

File: benchmarks/archive_stability.py [hunk 4]
```diff
@@ -0,0 +1,175 @@
...
+    return {
+        "stable": False,
+        "copies": len(paths),
+        "distinct_digests": len(by_digest),
+        "sizes": sizes,
+        "cause": cause,
+        "members_differing": len(differing) if differing is not None else None,
+        "example_members": (differing or [])[:5],
```
[the remark this answers is not in the document]
Comment (jdx) [reply benchmarks/archive_stability.py]: Already fixed in 66a0f21 (this review was against 8be417d) — `classify` now compares every distinct variant against a baseline and unions the differing members, so `timestamp` means all copies agree on every payload. On the sample cache it widened `libaws_lc_crypto.a` from 119 to 120 differing members, catching `__.SYMDEF SORTED`.

_🤖 Addressed by [Claude Code](https://claude.com/claude-code)_

**Comment by jdx** [comment]
Both findings in the latest review were already addressed in 66a0f21 — that review was made against 8be417d, the first commit.

On the Major one (effective archive policy in the build-script action key): fixed, though at a different location than suggested. Rather than folding it into `cargo_environment`, the `Action` struct carries a dedicated `archive_timestamps` field holding the `ZERO_AR_DATE` the script will actually read, including a value the developer set themselves — so two developers with different settings do not share an entry. The action version is bumped alongside it, since an entry stored earlier cannot say which policy produced it.

The specific scenario raised — the cache-hit path restoring before `apply_ar_determinism` configures the child — is covered: `build_action` is called at `build_script.rs:301`, before `restore()`, and computes the key from the environment directly rather than from the spawn. `test/archive_determinism.bats` flips the policy against a warm store to cover it, and that test was confirmed to fail when the key field is stubbed back to `None`.

_🤖 Addressed by [Claude Code](https://claude.com/claude-code)_

File: crates/mbx/src/build_script.rs [hunk 5]
```diff
@@ -220,6 +224,39 @@ fn install_launcher(mbx: &Path, executable: &Path) -> std::io::Result<()> {
...
+fn ar_profile() -> Option<String> {
+    std::env::var(crate::ar::PROFILE).ok()
+}
+
+/// The `ZERO_AR_DATE` this build script will see, for its cache key.
+fn archive_timestamp_key() -> Option<String> {
+    crate::ar::effective_zero_ar_date(
+        ar_mode(),
+        ar_profile().as_deref(),
+        std::env::var(crate::ar::ZERO_AR_DATE).ok(),
```
[the remark this answers is not in the document]
Comment (jdx) [reply crates/mbx/src/build_script.rs]: Fixed in ff2a468 — real bug, thanks. `archive_timestamp_key` now reads `var_os` like `apply_ar_determinism` does, and `effective_zero_ar_date` takes an `OsStr` so the value cannot be dropped on the way to the key. Values are spelled `text:<value>` or `bytes:<hex>` so a byte spelling cannot collide with the text that looks like it. Added unix-gated unit tests for a non-UTF-8 value keying distinctly from both absent and `"1"`.

_🤖 Addressed by [Claude Code](https://claude.com/claude-code)_

[the remark this answers is not in the document]
Comment (jdx) [reply crates/mbx/src/build_script.rs]: Already fixed in ff2a468 (this review was against 66a0f21). `archive_timestamp_key` now reads `var_os`, matching `apply_ar_determinism`, and `effective_zero_ar_date` takes an `OsStr` so the value cannot be dropped on the way to the key. Values are tagged `text:<value>` or `bytes:<hex>` so the three cases — missing, UTF-8, non-UTF-8 — stay distinct and a byte spelling cannot collide with the text that looks like it. Unix-gated tests cover a non-UTF-8 inherited value keying distinctly from both absent and `"1"`, and `bytes:c328` vs `text:c328`.

_🤖 Addressed by [Claude Code](https://claude.com/claude-code)_

File: benchmarks/archive_stability.py [hunk 6]
```diff
@@ -0,0 +1,281 @@
...
+            cause = "content" if substantive or cause == "content" else "build path"
+    if differing is not None:
+        differing = sorted(differing)
+    roots = build_roots(representatives) if cause in ("content", "build path") else []
+    distinct_roots = sorted({root for root in roots if root})
+    # Objects carrying different build directories explain a byte difference
+    # without anything being unstable, so say so rather than leaving "content"
+    # to be read as a defect.
+    if cause == "content" and len(distinct_roots) > 1:
+        cause = "build path"
```
[the remark this answers is not in the document]
Comment (jdx) [reply benchmarks/archive_stability.py]: Fixed in 4c6e6d0 — correct catch, and the override was mine from 0aa86f7. The cause is now decided once from what the comparison establishes, never overwritten: `content` when payloads still differ and the copies share a build directory (and always when the member sets disagree), and a new `undetermined` when payloads still differ but the copies were built at different directories — since a path that changes length also moves offsets inside an object, normalizing its text cannot prove the rest equal. That case now reports that it cannot tell, and that building at two equal-length paths would settle it, rather than calling it harmless. `libaws_lc_crypto.a` is `undetermined`; verified `build path` still holds for equal-length paths and `libz-ng.a` stays `timestamp`.

_🤖 Addressed by [Claude Code](https://claude.com/claude-code)_

[the remark this answers is not in the document]
Comment (jdx) [reply benchmarks/archive_stability.py:226]: Fixed in 4c6e6d0 — correct catch, and the override was mine from 0aa86f7. The cause is now decided once from what the comparison establishes, never overwritten: `content` when payloads still differ and the copies share a build directory (and always when the member sets disagree), and a new `undetermined` when payloads still differ but the copies were built at different directories — since a path that changes length also moves offsets inside an object, normalizing its text cannot prove the rest equal. That case now reports that it cannot tell, and that building at two equal-length paths would settle it, rather than calling it harmless. `libaws_lc_crypto.a` is `undetermined`; verified `build path` still holds for equal-length paths and `libz-ng.a` stays `timestamp`.

_🤖 Addressed by [Claude Code](https://claude.com/claude-code)_
